File: Docs/Algorithms/searching/binary_search.py
def binary_search(nums: List[int], target: int) -> int:
    """
    배열의 내부 데이터가 정렬되어 있어야만 사용할 수 있는 알고리즘
    TC O(logN): 확인하는 원소의 개수가 절반씩 줄어든다
    탐색 범위가 큰 상황에서의 탐색을 가정하는 문제가 많다.
    탐색 범위가 2000만을 넘어가면 이진 탐색을 고려해보자. (Python 에서 1초 내로 걸리려면 보통 2000만 번 이하, 2*10^7)
    """
    def bs_recursive(nums, start, end, target):
        if start > end:
            return -1
        mid = (start + end) // 2

        if nums[mid] == target:
            return mid
        elif nums[mid] > target:
            return bs_recursive(nums, start, mid - 1, target)
        else:
            return bs_recursive(nums, mid + 1, end, target)

        return -1

    def bs_stack(nums, target):
        stack = []

        while stack:
            start, end = stack.pop()
            if start > end:
                continue

            mid = (start + end) // 2
            if nums[mid] == target:
                return mid
            elif nums[mid] > target:
                stack.append((start, mid-1))
            else:
                stack.append((mid+1, end))

        return -1

File: Docs/Algorithms/searching/BinarySearch.java
package Docs.Algorithms.searching;

import java.util.Arrays;

public class BinarySearch {
}

class Example {
    static void main(String[] args) {
        int[] nums = { 1, 3, 5, 7, 9, 11 };
        int target = 7;

        int result = bSearch(nums, target);
        System.out.println("result idx: " + result);
    }

    static int bSearch(int[] nums, int target) {
        int left = 0;
        int right = nums.length - 1;

        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (nums[mid] == target) {
                return mid;
            } else if (nums[mid] < target) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }

        return -1;
    }
}

// ParametricSearch
class BudgetAllocation {
    static void main(String[] args) {
        int[] requests = { 120, 140, 110, 150 };
        int totalBudget = 485;

        int result = findMaxCap(requests, totalBudget);
        System.out.println("최대 상환액: " + result);
    }

    static int findMaxCap(int[] requests, int totalBudget) {
        int left = 0;
        int right = Arrays.stream(requests).max().getAsInt();

        while (left <= right) {
            int mid = left + (right - left) / 2;
            int allocated = calculateAllocated(requests, mid); // 상한액 후보

            if (allocated <= totalBudget) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        return right;
    }

    static int calculateAllocated(int[] requests, int cap) {
        int total = 0;
        for (int request : requests) {
            total += Math.min(request, cap); // 상한액 만큼만 배정
        }
        return total;
    }
}